### PR TITLE
Fix listeners

### PR DESCRIPTION
**Pull request by Jamdan2**
This just fixes some stuff about the listeners. First, it fixes an internal bug which would cause some lifecycle methods to get called at the wrong time. It also adds public api, so now any lifecycle can be added with either `add(lifecycle)` or `+lifecycle`. Lastly, instead of accepting only Subsystems, the Robot class's constructor now accepts all `RobotLifecycle` types.

### CODE REVIEW

File: core/src/test/kotlin/org/sertain/RobotTest.kt [hunk 1]
```diff
@@ -12,12 +12,12 @@ class RobotTest {
 
     @Before
     fun beforeEach() {
-        RobotLifecycle.addListener(testLifecycle)
+        +testLifecycle
```
Comment (andrewda) [inline core/src/test/kotlin/org/sertain/RobotTest.kt]: Gotta say, this is pretty lit

File: core/src/main/kotlin/org/sertain/Robot.kt [hunk 2]
```diff
@@ -205,3 +193,15 @@ public abstract class Robot(vararg subsystems: Subsystem) : IterativeRobot(), Ro
         DISABLED, AUTO, TELEOP
     }
 }
+
+public fun add(lifecycle: RobotLifecycle) {
```
Comment (andrewda) [inline core/src/main/kotlin/org/sertain/Robot.kt]: Add some docs please 😄 Just something like:

```kotlin
/**
 * @see RobotLifecycle.addListener
 */
```

or whatever the correct `@see` format is. Same thing for the three methods below.

**Comment by andrewda** [comment]
@Jamdan2 Did you get a chance to test this more on Thursday? Is it working?

**Comment by Jamdan2** [comment]
Tested it on thursday, it works 👌

**Pull request merged**